### PR TITLE
freetube: fix build, use latest electron

### LINKED ISSUE #264075: Build failure: Freetube
### Steps To Reproduce

1. build system configuration with Freetube in it.

### Build log

```
       … while evaluating derivation 'freetube-0.19.1'
         whose name attribute is located at /nix/store/bmcm18ngcq3pplprdkypa348qmi8lcps-nixos/nixos/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'postFixup' of derivation 'freetube-0.19.1'

         at /nix/store/bmcm18ngcq3pplprdkypa348qmi8lcps-nixos/nixos/pkgs/applications/video/freetube/default.nix:39:3:

           38|   # Electron version is set to 22 in order to match upstream
           39|   postFixup = ''
             |   ^
           40|     makeWrapper ${electron_22}/bin/electron $out/bin/${pname} \

       … from call site

         at /nix/store/bmcm18ngcq3pplprdkypa348qmi8lcps-nixos/nixos/pkgs/stdenv/generic/check-meta.nix:449:16:

          448|         {
          449|           no = handleEvalIssue { inherit meta attrs; } { inherit (validity) reason errormsg; };
             |                ^
          450|           warn = handleEvalWarning { inherit meta attrs; } { inherit (validity) reason errormsg; };

       … while calling 'handleEvalIssue'

         at /nix/store/bmcm18ngcq3pplprdkypa348qmi8lcps-nixos/nixos/pkgs/stdenv/generic/check-meta.nix:226:38:

          225|
          226|   handleEvalIssue = { meta, attrs }: { reason , errormsg ? "" }:
             |                                      ^
          227|     let

       error: Package ‘electron-22.3.27’ in /nix/store/bmcm18ngcq3pplprdkypa348qmi8lcps-nixos/nixos/pkgs/development/tools/electron/binary/generic.nix:35 is marked as insecure, refusing to evaluate.


       Known issues:
        - Electron version 22.3.27 is EOL

       You can install it anyway by allowing this package, using the
       following methods:

       a) To temporarily allow all insecure packages, you can use an environment
          variable for a single invocation of the nix tools:

            $ export NIXPKGS_ALLOW_INSECURE=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) for `nixos-rebuild` you can add ‘electron-22.3.27’ to
          `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
          like so:

            {
              nixpkgs.config.permittedInsecurePackages = [
                "electron-22.3.27"
              ];
            }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
          ‘electron-22.3.27’ to `permittedInsecurePackages` in
          ~/.config/nixpkgs/config.nix, like so:

            {
              permittedInsecurePackages = [
                "electron-22.3.27"
              ];
            }
```

### Additional context

See https://github.com/NixOS/nixpkgs/issues/263654 and https://github.com/NixOS/nixpkgs/pull/263685 for the exact same issue with Morgen.

Electron version lock seems to have been introduced with https://github.com/NixOS/nixpkgs/pull/252061 .

Upstream seems to be at Electron 27: https://github.com/FreeTubeApp/FreeTube/blob/development/yarn.lock#L3468 .

### Notify maintainers

@ryneeverett and @alyaeanyx 

### Metadata

```console
[tormod@nixos:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11pre537656.7c9cc5a6e5d3`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by kirillrdy**
## Description of changes
fixes https://github.com/NixOS/nixpkgs/issues/264075

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyaeanyx** [comment]
We originally pinned this to electron_22 to avoid mismatches with upstream, see https://github.com/NixOS/nixpkgs/pull/252061. Upstream is on 27.0.2 now, so could you pin this to `electron_27` instead of `electron` please?

**Comment by kirillrdy** [comment]
> We originally pinned this to electron_22 to avoid mismatches with upstream, see #252061. Upstream is on 27.0.2 now, so could you pin this to `electron_27` instead of `electron` please?

greetings !

I am not familiar enough with electron ecosystem. usually in nixpkgs, packages only pinned when necessary. Is there a nixos test we could write to check that app is at least not blank screen ? then we could just use latest electron

**Comment by alyaeanyx** [comment]
In this case, pinning electron is necessary sadly because of the breakages introduced by upstream matches.

Also i think a comment above `version = ` reminding people of bumping electron would also be a good idea, now that this part is moved out of the freetube default.nix

**Comment by ryneeverett** [comment]
It would be good to confirm that pinning 27 actually fixes the issue. I was always under the impression that electron-pinning would only be needed on a temporary basis.

**Comment by alyaeanyx** [comment]
> It would be good to confirm that pinning 27 actually fixes the issue.

Forgot to mention that, I already confirmed this fixes the issue but I didn't want to open a third PR

**Comment by kirillrdy** [comment]
> It would be good to confirm that pinning 27 actually fixes the issue

on master electron is 27. What issue pinning fixes ? ( apart from preventing possible future regressions )

**Comment by alyaeanyx** [comment]
Currently none, but as you alluded, future updates of electron on nixpkgs might introduce more breakages

**Comment by ryneeverett** [comment]
> Currently none, but as you alluded, future updates of electron on nixpkgs might introduce more breakages

I'm not sure I can agree with that rationale. A pinned version of electron is *guaranteed* to introduce breakages if it isn't updated. The pin to 27 could also easily break if 27 were marked insecure or removed from nixpkgs. Without automated pinning it doesn't seem worth it unless, per @kirillrdy's proposal, there were an automated test to verify that the pin has some value.

**Comment by alyaeanyx** [comment]
My idea was that every time a new freetube version comes out, we check which electron version was packaged with it and change the pin in nixpkgs accordingly. I agree that _27 being marked as insecure/etc could be a problem though - it's a tradeoff, and a rather uncertain one, because we don't know what further breakages might be introduced with future versions. So yeah, let's remove the pin for now and add it in again in case an electron version mismatch is causing problems in the future.

**Comment by kirillrdy** [comment]
@alyaeanyx @ryneeverett I've written a very basic nixosTest, let me know if's sufficient

**Comment by ryneeverett** [comment]
That does not currently work as a regression test because it succeeds on the parent commit of #252061:

```sh
git checkout c236073d10820c7147f89c3dd2dd5de161c79136
git cherry-pick -n aa06a9c77a5bbc45520875523539c60096c3e7d2
nix-build -A pkgs.freetube.passthru.tests
```

Looking at the test it makes sense because you'd need to run it on a wayland VM to reproduce the bug and it appears to be X.

**Comment by kirillrdy** [comment]
> That does not currently work as a regression test because it succeeds on the parent commit of #252061:
> 
> ```shell
> git checkout c236073d10820c7147f89c3dd2dd5de161c79136
> git cherry-pick -n aa06a9c77a5bbc45520875523539c60096c3e7d2
> nix-build -A pkgs.freetube.passthru.tests
> ```
> 
> Looking at the test it makes sense because you'd need to run it on a wayland VM to reproduce the bug and it appears to be X.

fair point !
i'll adjust test so that run against both xorg and wayland

**Comment by kirillrdy** [comment]
@ryneeverett I've pushed wayland and xorg tests, have a look

**Comment by ryneeverett** [comment]
How do you run the tests now that the file is an attrset?

**Comment by kirillrdy** [comment]
> How do you run the tests now that the file is an attrset?

`nix-build -A freetube.passthru.tests` or `nix-build -A nixosTests.freetube`

`passthru.tests` was always attribute set ( even with 1 test )

`nix-build` can build attribute sets, it will just build all derivations, so you will get multiple `result` symlinks

**Comment by kirillrdy** [comment]
you can also selectively just run 1 test eg

```
nix-build -A freetube.passthru.tests.freetube.xorg
```

File: pkgs/applications/video/freetube/default.nix [hunk 1]
```diff
@@ -9,6 +9,10 @@ stdenv.mkDerivation rec {
     sha256 = "add96ad3509d4d5c6d8658b005dfd046963cd6bb0a4e1f3e88f726a86c05810f";
   };
 
+  passthru = {
+    tests = { inherit (nixosTests) freetube; };
```
Comment (kirillrdy) [inline pkgs/applications/video/freetube/default.nix:13]: actually, i think I will remove unnecessary nesting

**Comment by ryneeverett** [comment]
Thanks. The test still doesn't seem to work as a regression test because it still passes on wayland at c236073d10820c7147f89c3dd2dd5de161c79136. Then again, I was never able to reproduce this bug on wayland in the first place.

**Comment by kirillrdy** [comment]
> Thanks. The test still doesn't seem to work as a regression test because it still passes on wayland at [c236073](https://github.com/NixOS/nixpkgs/commit/c236073d10820c7147f89c3dd2dd5de161c79136). Then again, I was never able to reproduce this bug on wayland in the first place.

@ryneeverett actually, I can reproduce the issue on my wayland laptop setup, will try to update test later today

**Comment by kirillrdy** [comment]
@ryneeverett latest version, catches regression at c236073d10820c7147f89c3dd2dd5de161c79136

the only annoying thing right now, wait_for_text has 15 minutes timeout. To fix that will require PR to nixpkgs test-driver code, for now I think its ok, and its better than adding a sleep in the test.

**Comment by ryneeverett** [comment]
Hmm, it still passes at c236073d10820c7147f89c3dd2dd5de161c79136 on my machine. But if folks who have reproduced this bug on metal are satisfied that's good enough for me.

**Comment by kirillrdy** [comment]
> Hmm, it still passes at [c236073](https://github.com/NixOS/nixpkgs/commit/c236073d10820c7147f89c3dd2dd5de161c79136) on my machine. But if folks who have reproduced this bug on metal are satisfied that's good enough for me.

@ryneeverett 

I can reproduce on both of my machines, 11th and 13th gen intel i7, although tests run in qemu, it's still possible that its timing related, but just in case ( just so that we are sure that we are running same thing )

```
git clone https://github.com/kirillrdy/nixpkgs -b broken --depth 1 broken
cd broken
nix-build -A nixosTests.freetube.wayland
```

also, ultimate test, we should see if we can trigger borg to fail

**Comment by a-kenji** [comment]
@kirillrdy 
> the only annoying thing right now, wait_for_text has 15 minutes timeout. To fix that will require PR to nixpkgs test-driver code, for now I think its ok, and its better than adding a sleep in the test.

The `wait_for` functions allow configuring the timeout:
image

https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests

**Comment by kirillrdy** [comment]
> @kirillrdy
> 
> > the only annoying thing right now, wait_for_text has 15 minutes timeout. To fix that will require PR to nixpkgs test-driver code, for now I think its ok, and its better than adding a sleep in the test.
> 
> The `wait_for` functions allow configuring the timeout: image
> 
> https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests

`wait_for_text` doesn't have timeout

**Comment by a-kenji** [comment]
Are you sure? The autocompletion shows a timeout parameter for me:
image

**Comment by kirillrdy** [comment]
> @kirillrdy
> 
> > the only annoying thing right now, wait_for_text has 15 minutes timeout. To fix that will require PR to nixpkgs test-driver code, for now I think its ok, and its better than adding a sleep in the test.
> 
> The `wait_for` functions allow configuring the timeout: image
> 
> https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests

@a-kenji correction wait_for_text has timeout, just not at the commit where were testing breakage

thanks you for making me check that

**Comment by ryneeverett** [comment]
> I can reproduce on both of my machines, 11th and 13th gen intel i7, although tests run in qemu, it's still possible that its timing related, but just in case ( just so that we are sure that we are running same thing )
> 
> ```
> git clone https://github.com/kirillrdy/nixpkgs -b broken --depth 1 broken
> cd broken
> nix-build -A nixosTests.freetube.wayland
> ```

Still succeeds on my AMD fx-8320.

I wouldn't worry too much about it though. Your tests are valuable even if they wouldn't catch that particular bug. My main concern was that we have a proper regression test to justify pinning the electron version. You're not doing that any more so that concern is moot.

AFAIK, we don't have any reason to believe that this bug will recur or that if it does it will be caused by an electron version-mismatch. My impression is that many applications in nixpkgs get by without pinning electron version so hopefully this issue was anomalous.

**Review by ryneeverett — APPROVED** [review]

**Comment by kirillrdy** [comment]
ok, so looks like ofborg on aarch64-linux timed out, not sure what to do there, apart from retry

as for the failure, the issue seems pretty straight forward, may I nag you one last time @ryneeverett to try `nix-build -A nixosTests.freetube.wayland --pure`

otherwise we'll just wait for ofborg to pass

**Comment by kirillrdy** [comment]
looks like for vscodium ( where I took inspiration for nixosTests ) aarch64-linux fails as well with similar timeout error

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 264084` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 test built:</summary>
  <ul>
    <li>nixosTests.freetube.wayland</li>
  </ul>
</details>

**Review by natsukium — APPROVED** [review]
LGTM

**Comment by kirillrdy** [comment]
Thank you very much @natsukium san

**Pull request merged**